internal/benchrunner/runners/rally/metrics.go

```diff
@@ -0,0 +1,370 @@
+}
+
+type metrics struct {
+	ts        int64
+	dsMetrics *ingest.DataStreamStats
+	nMetrics  *ingest.NodesStats
+}
+
+type metricsSummary struct {
+	ClusterName         string
+	Nodes               int
+	RunID               string
+	CollectionStartTs   int64
+	CollectionEndTs     int64
+	DataStreamStats     *ingest.DataStreamStats
+	IngestPipelineStats map[string]ingest.PipelineStatsMap
+	DiskUsage           map[string]ingest.DiskUsage
+	TotalHits           int
+	NodesStats          map[string]ingest.NodeStats
+}
+
+func newCollector(
+	ctxt servicedeployer.ServiceContext,
+	benchName string,
+	scenario scenario,
+	esAPI, metricsAPI *elasticsearch.API,
+	interval time.Duration,
+	datastream, pipelinePrefix string,
+) *collector {
+	meta := benchMeta{Parameters: scenario}
+	meta.Info.Benchmark = benchName
+	meta.Info.RunID = ctxt.Test.RunID
+	return &collector{
+		ctxt:           ctxt,
+		interval:       interval,
+		scenario:       scenario,
+		metadata:       meta,
+		esAPI:          esAPI,
+		metricsAPI:     metricsAPI,
+		datastream:     datastream,
+		pipelinePrefix: pipelinePrefix,
+		stopC:          make(chan struct{}),
+	}
+}
+
+func (c *collector) start() {
+	c.tick = time.NewTicker(c.interval)
+	c.createMetricsIndex()
+	var once sync.Once
+
+	c.wg.Add(1)
+	go func() {
+		defer c.tick.Stop()
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.stopC:
+				// last collect before stopping
+				c.collectMetricsPreviousToStop()
+				c.publish(c.createEventsFromMetrics(c.endMetrics))
+				return
+			case <-c.tick.C:
+				once.Do(func() {
+					c.waitUntilReady()
+					c.startIngestMetrics = c.collectIngestMetrics()
+					c.startTotalHits = c.collectTotalHits()
+					c.startMetrics = c.collect()
+					c.publish(c.createEventsFromMetrics(c.startMetrics))
+				})
+				m := c.collect()
+				c.publish(c.createEventsFromMetrics(m))
+			}
+		}
+	}()
+}
+
+func (c *collector) stop() {
+	if !c.stopped.CompareAndSwap(false, true) {
+		return
+	}
+	close(c.stopC)
+	c.wg.Wait()
+}
+
+func (c *collector) collect() metrics {
+	m := metrics{
+		ts: time.Now().Unix(),
+	}
+
+	nstats, err := ingest.GetNodesStats(c.esAPI)
+	if err != nil {
+		logger.Debug(err)
+	} else {
+		m.nMetrics = nstats
+	}
+
+	dsstats, err := ingest.GetDataStreamStats(c.esAPI, c.datastream)
+	if err != nil {
+		logger.Debug(err)
+	} else {
+		m.dsMetrics = dsstats
+	}
+
+	return m
+}
+
+func (c *collector) publish(events [][]byte) {
+	if c.metricsAPI == nil {
+		return
+	}
+	for _, e := range events {
+		reqBody := bytes.NewReader(e)
+		resp, err := c.metricsAPI.Index(c.indexName(), reqBody)
+		if err != nil {
+			logger.Debugf("error indexing event: %v", err)
+			continue
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			logger.Errorf("failed to read index response body: %v", err)
+		}
+		resp.Body.Close()
+
+		if resp.StatusCode != 201 {
+			logger.Errorf("error indexing event (%d): %s: %v", resp.StatusCode, resp.Status(), elasticsearch.NewError(body))
```

Comment:
it might be a transient error.
to my understanding publishing events to the metricsstore is best effort: we should not stop the benchmark if it fails
but we do log the error, so the user is aware of that